gnomad_qc/v4/sample_qc/create_sample_qc_metadata_ht.py

```diff
@@ -437,35 +433,117 @@ def _sample_check(
         global_ann = {f"{ann_label}_parameters": ann_ht.index_globals()}
 
-    ht = ht.annotate(**ann)
+    ht = base_ht.annotate(**ann)
     ht = ht.annotate_globals(**global_ann)
+
     return ht
 
 
-def main(args):
-    """Merge the output of all sample QC modules into a single Table."""
-    hl.init(
-        log="/sample_metadata.log",
-        default_reference="GRCh38",
-        tmp_dir="gs://gnomad-tmp-4day",
+def get_sample_filter_ht(base_ht: hl.Table, relationship_ht: hl.Table) -> hl.Table:
+    """
+    Combine sample filters into a single Table to be added to the metadata Table.
+
+    Includes hard-filters, sample QC outlier filters, and relatedness filters.
+
+    :param base_ht: Input Table to add annotations to.
+    :param relationship_ht: Table with relationships annotations.
+    :return: Table with hard filter metric annotations added.
+    """
+    logger.info("Combining sample filters Tables for 'sample_filters' struct.")
+    # Get list of UKB samples that should be removed.
+    ukb_remove = hl.import_table(all_ukb_samples_to_remove, no_header=True).f0.collect()
+    # Get list of hard filtered samples with sex imputation.
+    hf_s = hard_filtered_samples.ht().s.collect()
+
+    hard_filters_ht = get_hard_filters_ht(base_ht)
+    outlier_filters_ht = finalized_outlier_filtering().ht()
+    relatedness_filters_ht = annotate_relatedness_filters(
+        ht=base_ht,
+        relationship_ht=relationship_ht,
+        hard_filtered_expr=hard_filters_ht[base_ht.key].hard_filtered,
+        outlier_filtered_expr=outlier_filters_ht[base_ht.key].outlier_filtered,
+    )
+    sample_filters = {
+        "hard_filters": {
+            "ann_ht": hard_filters_ht,
+            "global_top_level": True,
+        },
+        "outlier_detection": {
+            "ann_ht": outlier_filters_ht,
+            "base_ht_missing": ukb_remove,
+            "ann_ht_missing": hf_s,
+        },
+        "relatedness_filters": {
+            "ann_ht": relatedness_filters_ht,
+            "global_top_level": True,
+        },
+    }
+
+    sample_filters = [
+        {**ann_params, **{"ann_label": ann, "ann_top_level": True}}
+        for ann, ann_params in sample_filters.items()
+    ]
+    sample_filters_ht = reduce(
+        lambda ht, ann_params: add_annotations(ht, **ann_params),
+        [base_ht] + sample_filters,
     )
+    sample_filters_ht = sample_filters_ht.checkpoint(
+        new_temp_file("sample_filters", extension="ht"), overwrite=True
+    )
+
+    return sample_filters_ht
+
+
+def get_hard_filter_metric_ht(base_ht: hl.Table) -> hl.Table:
+    """
+    Combine sample contamination, chr20 mean DP, and QC MT callrate into a single Table.
+
+    :param base_ht: Input Table to add annotations to.
+    :return: Table with hard filter metric annotations added.
+    """
+    logger.info("Combining hard-filter metric Tables for 'hard_filter_metrics' struct.")
+
+    # TODO: Add drop of `gq_thresholds` to the sample_chr20_mean_dp code.
```

Comment:
Post v4 -- but what are your thoughts on creating issues for all the TODOs that are going to be merged to main and adding the ticket number to the TODO? I think this could help with getting to the TODO done but obviously is not a priority, just a thought on workflow/organization.

Reply:
Yeah that's a great idea, I will do that for the TODO's that we want for this script. For this specific TODO I might change this to a note that it wasn't dropped, and make the ones in the actual files sound more like a warning. I didn't want to actually change the v4 code since the current files will still have them, but wanted to warn others if they copy or modify the code later.